deps/thpool/thpool.c

```diff
@@ -437,17 +432,92 @@ static struct job* jobqueue_pull(jobqueue* jobqueue_p) {
   jobqueue_clear(jobqueue_p);
-  rm_free(jobqueue_p->has_jobs);
+}
+
+/* ======================== PRIORITY QUEUE ========================== */
+
+static int priority_queue_init(priority_queue* priority_queue_p) {
+  
+  priority_queue_p->has_jobs = (struct bsem*)rm_malloc(sizeof(struct bsem));
+  if (priority_queue_p->has_jobs == NULL) {
+    return -1;
+  }
+  jobqueue_init(&priority_queue_p->high_priority_jobqueue);
+  jobqueue_init(&priority_queue_p->low_priority_jobqueue);
+  bsem_init(priority_queue_p->has_jobs, 0);
+  pthread_mutex_init(&priority_queue_p->jobqueues_rwmutex, NULL);
+  priority_queue_p->pulls = 0;
+  return 0;
+}
+
+static void priority_queue_clear(priority_queue* priority_queue_p) {
+  jobqueue_clear(&priority_queue_p->high_priority_jobqueue);
+  jobqueue_clear(&priority_queue_p->low_priority_jobqueue);
+  bsem_reset(priority_queue_p->has_jobs);
+}
+
+static void priority_queue_push(priority_queue* priority_queue_p, struct job* newjob_p, thpool_priority priority) {
+  pthread_mutex_lock(&priority_queue_p->jobqueues_rwmutex);
+  switch (priority) {
+    case THPOOL_PRIORITY_HIGH:
+      jobqueue_push(&priority_queue_p->high_priority_jobqueue, newjob_p);
+      break;
+    case THPOOL_PRIORITY_LOW:
+      jobqueue_push(&priority_queue_p->low_priority_jobqueue, newjob_p);
+      break;
+  }
+  bsem_post(priority_queue_p->has_jobs);
+  pthread_mutex_unlock(&priority_queue_p->jobqueues_rwmutex);
+}
+
+static struct job* priority_queue_pull(priority_queue* priority_queue_p) {
+  struct job* job_p = NULL;
+  pthread_mutex_lock(&priority_queue_p->jobqueues_rwmutex);
+  // We want to pull from the lower priority queue every 3rd time
+  if (priority_queue_p->pulls % 3 == 2) {
+    job_p = jobqueue_pull(&priority_queue_p->low_priority_jobqueue);
+    // If the lower priority queue is empty, pull from the higher priority queue
+    if(!job_p) {
+      job_p = jobqueue_pull(&priority_queue_p->high_priority_jobqueue);
+    }
+  } else {
+    job_p = jobqueue_pull(&priority_queue_p->high_priority_jobqueue);
+    // If the higher priority queue is empty, pull from the lower priority queue
+    if(!job_p) {
+      job_p = jobqueue_pull(&priority_queue_p->low_priority_jobqueue);
+    }
+  }
+  priority_queue_p->pulls++;
+  if(priority_queue_p->high_priority_jobqueue.len ||  priority_queue_p->low_priority_jobqueue.len ) {
+    bsem_post(priority_queue_p->has_jobs);
+  }
+  pthread_mutex_unlock(&priority_queue_p->jobqueues_rwmutex);
+
+  return job_p;
+
+}
+
+static void priority_queue_destroy(priority_queue* priority_queue_p) {
+  jobqueue_destroy(&priority_queue_p->high_priority_jobqueue);
+  jobqueue_destroy(&priority_queue_p->low_priority_jobqueue);
+  rm_free(priority_queue_p->has_jobs);
+  pthread_mutex_destroy(&priority_queue_p->jobqueues_rwmutex);
+}
+
+static size_t priority_queue_len(priority_queue* priority_queue_p) {
+  size_t len;
+  pthread_mutex_lock(&priority_queue_p->jobqueues_rwmutex);
```

Comment:
`thpool_p->thcount_lock` is held by the caller, isn't it the guard of `len`? Why should we hold `priority_queue_p->jobqueues_rwmutex` here as well?